### PR TITLE
Redo query column/table relationship without TableReferenceExpression 

### LINKED ISSUE #32558: Look into making SelectExpression (more) immutable, getting rid of TableReferenceExpression
Our ColumnExpressions currently reference their tables via TableReferenceExpressions, which are themselves mutable; this is the source of many bugs, e.g. when the SelectExpression is copied (e.g. when a visitor changes something inside), but the columns remain shared between the two copies, causing various referential integrity problems (e.g. #32234). This system also adds substantial complexity to keep everything in sync, again adding a source of bugs.

To simplify this, we'll consider removing TableReferenceExpression altogether, and have ColumnExpression reference its TableExpressionBase via its string alias only; there would be no actual .NET reference  between the column and the table.

* This makes the SQL tree **almost** immutable
    * SelectExpression still has a temporary mutable mode during translation - this is orthogonal to the table/column relationship and their mutability.
    * At the moment, the table alias is a property on TableExpressionBase; there's no easy way to immutably change that alias - it's a mutable property. We need to make that immutable as well.
* Any visitor that needs the actual TableExpressionBase given a ColumnExpression would need to track aliases/tables as part of its state and perform lookups.
    * This effectively moves the referential relationship between column and table out of the expression tree, and into visitor state, which is temporary/point-in-time.
    * There are only a few such actual needs; the type mapping inference visitor is one, and the pruner makes an optional use of this as well. In any case, the additional work for tracking alias/table mappings is negligible and does not add any considerable complexity.
* Once we're fully immutable, we can consider removing expression cloning altogether. We may still decide to continue doing it only so that cloned fragments inside the same tree have different table aliases (though note that we've never had fully unique aliases in the tree). Note that we'd ideally stop having cloned duplicates within the same query tree (i.e. GroupBy), at which point this will become totally moot. 
* Note that this depends on a change in how we manage aliases (see [#32784](https://github.com/dotnet/efcore/issues/32784)).
    * Currently, we generate temporary aliases for tables created in SQL fragments, and then run a uniquifying visitor when that fragment is added to a SelectExpression in the tree (predicate, ordering...).
    * This "late" alias uniquification would require us to also replace all columns referencing the table (and is generally inefficient even without that).
    * Instead, we should generate a unique alias up-front, and pass it to the table when it's created, not changing it later (except for post-processing, at which point we'd need a pass to rewrite the referencing columns).

Note that we considered two alternatives:

* Have ColumnExpression directly reference its TableExpressionBase (without TableReferenceExpression).
    * This is also a fully immutable model; but it means that any change to a TableExpressionBase requires a additional recursive visit to all the SelectExpression's clauses to replace any ColumnExpressions (and their containing expressions). Recall that a table can be a SELECT subquery, at which point any modification within would trigger such a recursive rewrite. This is quite a heavy alternative.
    * Referencing the table via its alias is a bit similar, but only requires recursive rewriting when the alias changes, and otherwies allows effort-less replacing of tables. The only visitor which currently modifies aliases is the alias manager post-processor (which closes alias gaps due to e.g. truncated tables); this seems very acceptable (and the postprocessing step is also completely optional/esthetic).
* Keep TableReferenceExpression, but make it owned by TableExpressionBase rather than the SelectExpression
    * Each TableExpressionBase would reference its TableReferenceExpression.
    * When a table needs to be replaced, we go to the old table's TRE and set it to point to the new table. Crucially, the new table must also share the same TRE from the moment it's created, to ensure that we don't have two TREs.
    * This would resolve a lot of the book-keeping and complexity concerns associated with maintaining a list of TableReferenceExpressions on SelectExpression, mirroring the list of tables.
    * This keeps the SQL tree mutable, meaning that we need to clone to reuse, etc.

Note that this has nothing to do with the SelectExpression having a "mutable mode", where its projection hasn't yet been applied (that's completely orthogonal to this).

### PR DESCRIPTION
**Pull request by roji**
~**NOTE: THIS IS BASED ON TOP OF #32785 (ALIAS REDO), REVIEW ONLY THE 2ND COMMIT**~

* After this, TableReferenceExpression no longer exists, and ColumnExpressions refer to their tables by their string aliases.
* Lots of table reference book-keeping/maintenance code has been removed, including various visitation passes (compilation performance).
* There aren't many places where a visitor actually needs to resolve a column to its table; in most cases these are on the current/top-most SelectExpression, which is trivial. In cases where columns can refer up the SQL tree, the visitor tracks the mapping as part of its visitation (this is very rare).
* ColumnExpression is now a regular, publicly-constructible expression - there's no more private ConcreteColumnExpression.
* Note that after this PR, table aliases are still mutable - a small upcoming PR will finish that bit off (in the interest of not cramming too much into a single PR).

Closes #32558

### CODE REVIEW

File: src/EFCore.Relational/Query/RelationalQueryableMethodTranslatingExpressionVisitor.cs [hunk 1]
```diff
@@ -1709,23 +1709,20 @@ Expression ExpandOwnedNavigation(INavigation navigation)
...
 
-            static TableExpressionBase FindRootTableExpressionForColumn(ColumnExpression column)
+            static TableExpressionBase FindRootTableExpressionForColumn(SelectExpression select, ColumnExpression column)
             {
-                var table = column.Table;
-                if (table is JoinExpressionBase joinExpressionBase)
-                {
-                    table = joinExpressionBase.Table;
-                }
-                else if (table is SetOperationBase setOperationBase)
```
Comment (roji) [inline src/EFCore.Relational/Query/RelationalQueryableMethodTranslatingExpressionVisitor.cs:1713]: @maumar unless I'm mistaken, there was a bug here when a set operation was wrapped in a join; we'd unwrap the join above but not go into the set operation because of the "else" here.

File: src/EFCore.Relational/Query/RelationalQueryableMethodTranslatingExpressionVisitor.cs [hunk 2]
```diff
@@ -2139,6 +2144,11 @@ public IReadOnlyDictionary<(TableExpressionBase, string), RelationalTypeMapping?
 
         protected override Expression VisitExtension(Expression node)
         {
+            if (node is TableExpressionBase { Alias: string tableAlias } table)
```
Comment (roji) [inline src/EFCore.Relational/Query/RelationalQueryableMethodTranslatingExpressionVisitor.cs:2147]: This is one of the few examples of a visitor which really does need to go from columns to tables, and so maintains a table map during visitation.

File: src/EFCore.Relational/Query/SqlAliasManager.cs [hunk 3]
```diff
@@ -181,21 +180,28 @@ protected override Expression VisitExtension(Expression node)
...
 
-                    return base.VisitExtension(node);
+                    if (aliasRewritingMap.TryGetValue(alias, out var newAlias))
+                    {
+                        // TODO: TableExpressionBase is still mutable with regards to its alias - this needs to change.
+                        // TODO: This visitor needs to replace the table in the usual way; but we don't currently have a good way of
+                        // TODO: recreating a new TableExpressionBase implementation with a different alias.
+                        // TODO: Either add a mandatory, abstract non-destructive ChangeAlias to TableExpressionBase, or ideally,
+                        // TODO: refactor things to split the alias out of TableExpressionBase altogether.
+                        table.Alias = newAlias;
```
Comment (roji) [inline src/EFCore.Relational/Query/SqlAliasManager.cs:198]: Note that TableExpressionBase.Alias is still mutable. It will be made immutable in a separate PR.

File: src/EFCore.Relational/Query/SqlExpressions/ColumnExpression.cs [hunk 4]
```diff
@@ -13,55 +13,85 @@ namespace Microsoft.EntityFrameworkCore.Query.SqlExpressions;
 ///     </para>
 /// </summary>
 [DebuggerDisplay("{TableAlias}.{Name}")]
-public abstract class ColumnExpression : SqlExpression
+public class ColumnExpression : SqlExpression
```
Comment (roji) [inline src/EFCore.Relational/Query/SqlExpressions/ColumnExpression.cs:16]: Note that ColumnExpression is no longer abstract, with a private ConcreteColumnExpression that only SelectExpression can construct. This is now a plain, regular SqlExpression that anyone can build.

File: src/EFCore.Relational/Query/SqlExpressions/SelectExpression.cs [hunk 5]
```diff
@@ -4661,57 +4566,56 @@ private static IEnumerable<INavigation> GetAllNavigationsInHierarchy(IEntityType
...
 
     /// <inheritdoc />
     protected override Expression VisitChildren(ExpressionVisitor visitor)
-        => VisitChildren(visitor, updateColumns: true);
-
-    private Expression VisitChildren(ExpressionVisitor visitor, bool updateColumns)
     {
         if (_mutable)
         {
+            VisitList(_tables, inPlace: true, out _);
```
Comment (roji) [inline src/EFCore.Relational/Query/SqlExpressions/SelectExpression.cs:4617]: Note that table visitation has been moved to be the first. This aligns with actual evaluation order of SQL SELECT - tables are evaluated first, and projections last - they can reference tables (the textual layout of SQL SELECT has always seemed like a mistake; IDEs can't to Intellisense as the user is writing projections since the tables haven't been written yet... Note that in C# LINQ expression syntax the tables come first).

This is a more useful for visitors which need to visit tables first, for tracking purposes etc.

File: src/EFCore.Relational/Query/SqlExpressions/SelectExpression.cs [hunk 6]
```diff
@@ -4895,39 +4791,6 @@ private Expression VisitChildren(ExpressionVisitor visitor, bool updateColumns)
...
 
-                // We duplicated the SelectExpression, and must therefore also update all table reference expressions to point to it.
-                // If any tables have changed, we must duplicate the TableReferenceExpressions and replace all ColumnExpressions to use
-                // them; otherwise we end up two SelectExpressions sharing the same TableReferenceExpression instance, and if that's later
-                // mutated, both SelectExpressions are affected (this happened in AliasUniquifier, see #32234).
-
-                // Otherwise, if no tables have changed, we mutate the TableReferenceExpressions (this was the previous code, left it for
-                // a more low-risk fix). Note that updateColumns is false only if we're already being called from
-                // ColumnTableReferenceUpdater to replace the ColumnExpressions, in which case we avoid infinite recursion.
-                if (tablesChanged && updateColumns)
```
Comment (roji) [inline src/EFCore.Relational/Query/SqlExpressions/SelectExpression.cs:4914]: All this table reference updating/cloning is now gone 🎉

File: src/EFCore.Relational/Query/SqlTreePruner.cs [hunk 7]
```diff
@@ -16,13 +16,29 @@ namespace Microsoft.EntityFrameworkCore.Query;
...
+    /// </summary>
+    /// <remarks>
+    ///     This is an internal API that supports the Entity Framework Core infrastructure and not subject to
+    ///     the same compatibility standards as public APIs. It may be changed or removed without notice in
+    ///     any release. You should only use it directly in your code with extreme caution and knowing that
+    ///     doing so can result in application failures when updating to a new Entity Framework Core release.
+    /// </remarks>
+    // TODO: Make this protected after SelectExpression.Prune is moved into this visitor
+    [EntityFrameworkInternal]
+    public virtual string? CurrentTableAlias { get; set; }
```
Comment (roji) [inline src/EFCore.Relational/Query/SqlTreePruner.cs:41]: When everything is done, I plan to do a final cleanup here to move the SelectExpression.Prune functionailty into this visitor; at that point everything will be nice and self-contained. Right now it's somewhat smelly - that's a temporary situation.

File: src/EFCore.Relational/Query/SqlExpressions/TableExpressionBase.cs [hunk 8]
```diff
@@ -149,4 +149,14 @@ public virtual TableExpressionBase AddAnnotation(string name, object? value)
...
+
+    /// <summary>
+    ///     This is an internal API that supports the Entity Framework Core infrastructure and not subject to
+    ///     the same compatibility standards as public APIs. It may be changed or removed without notice in
+    ///     any release. You should only use it directly in your code with extreme caution and knowing that
+    ///     doing so can result in application failures when updating to a new Entity Framework Core release.
+    /// </summary>
+    [EntityFrameworkInternal]
+    public virtual string GetRequiredAlias()
+        => Alias ?? throw new InvalidOperationException($"No alias is defined on table: {ExpressionPrinter.Print(this)}");
```
Comment (maumar) [inline src/EFCore.Relational/Query/SqlExpressions/TableExpressionBase.cs:161]: add resource string

File: src/EFCore.Relational/Query/RelationalQueryableMethodTranslatingExpressionVisitor.cs [hunk 9]
```diff
@@ -2128,14 +2125,22 @@ private bool TryExtractBareInlineCollectionValues(ShapedQueryExpression shapedQu
...
-        private readonly Dictionary<(TableExpressionBase, string), RelationalTypeMapping?> _inferredColumns = new();
+        private readonly Dictionary<(string TableAlias, string ColumnName), RelationalTypeMapping?> _inferredColumns = new();
 
-        private SelectExpression? _currentSelectExpression;
+        /// <summary>
+        ///     A mapping of table aliases to the <see cref="TableExpressionBase" /> instances; these are used to check the table type
+        ///     when we encounter a typed column pointing to it, and avoid recording inferred type mappings where we know the table
+        ///     doesn't need to be inferred from the column.
+        /// </summary>
+        private readonly Dictionary<string, TableExpressionBase> _tables = new();
```
Comment (maumar) [inline src/EFCore.Relational/Query/RelationalQueryableMethodTranslatingExpressionVisitor.cs:2135]: nit: maybe change name to something like _tableAliasMap? That's the convention I have been using at least to indicate dictionary rather than a simple list

**Review by AndriySvyryd — APPROVED** [review]

**Review by maumar — APPROVED** [review]

File: src/EFCore.Relational/Query/SqlExpressions/TableExpressionBase.cs (hunk 8 above)
Comment (roji) [reply src/EFCore.Relational/Query/SqlExpressions/TableExpressionBase.cs:161]: Oops missed this one, will do in the #32815

Comment (roji) [reply src/EFCore.Relational/Query/SqlExpressions/TableExpressionBase.cs:161]: Oops missed this one, will do in the #32815
